iso.bbx

```diff
@@ -543,17 +543,22 @@
 % Macro for availability and access information
 % with order of importance: doi, eprint, url
 \newbibmacro*{availability+access}{%
-  \iffieldundef{doi}
-    {\iffieldundef{eprint}
-       {\iftoggle{bbx:url}
-         {\printfield{url}}
-         {}}
-       {\iftoggle{bbx:eprint}
-         {\usebibmacro{from-eprint}}
-         {}}}
-    {\iftoggle{bbx:doi}
-      {\usebibmacro{from-doi}}
-      {}}%
+  \ifboolexpr{
+    test {\iftoggle{bbx:doi}}
+    and not test {\iffieldundef{doi}}
+  }
+    {\usebibmacro{from-doi}}%
+    {%
+      \ifboolexpr{
+        test {\iftoggle{bbx:eprint}}
+        and not test {\iffieldundef{eprint}}
+      }
+        {\usebibmacro{from-eprint}}%
+        {%
+          \iftoggle{bbx:url}
+            {\printfield{url}}%
```

Comment:
I'm wondering if `urldate` should also be controlled with a similar structure or should live here in the first place. (In fact `urldate` can be quite a large distance apart from this macro. In the `@thesis` driver for example the supervisor info is squeezed in between them when `bbx:thesisinfoinnotes` is false.)

In
```latex
\documentclass[english,british]{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{babel}
\usepackage{csquotes}

\usepackage[style=iso-authoryear, url=false, backend=biber]{biblatex}


\begin{filecontents}[force]{\jobname.bib}
@online{appleby,
  author  = {Humphrey Appleby},
  title   = {On the Importance of the Civil Service},
  date    = {1980-05-06},
  url     = {http://example.com/~sir_humphrey/imp.pdf},
  urldate = {2020-01-15},
}
\end{filecontents}
\addbibresource{\jobname.bib}
\addbibresource{biblatex-examples.bib}


\begin{document}
\cite{markey,appleby}
\printbibliography
\end{document}
```
the `urldate` on its own looks a bit weird.

The standard styles ignore the setting of the option `url` for `@online` entries so that the URL is always shown for `@online` entries.

Reply:
It makes sense that if there is no url printed, information when it has been seen is redundant. The same applies also for a medium type, like `[online]` as a `howpublished` field.
I am not really keen on hiding information that are available, since they may help to identify the resource, but of course, sometimes they're not necessary to clearly identify the material (as the standard requires) and might be dropped out.

One issue I can see is an online article, for instance, that might be declared in bib file as `@article` entry with `howpublished = {online article}` or similar and it will be weird to see `[online article]` and no url in the bibliography. That's what I dealt with in eb193e8e0f06af5344918412b080b3dd3ef8be04

BTW, from where is the `force` option in
```latex
\begin{filecontents}[force]{\jobname.bib}
```
because it does not compile in my standard TeX Live 2019.

Reply:
The `[force]` option was added in LaTeX 2019-10-01 (https://www.latex-project.org/news/latex2e-news/ltnews30.pdf) as a kernel replacement of the functions of the `filecontents` package.

Reply:
The following MWE shows what I meant above with the `urldate` and `url` being unreasonably far apart

```latex
\documentclass[british]{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{babel}
\usepackage{csquotes}

\usepackage[style=iso-authoryear, thesisinfoinnotes=false, backend=biber]{biblatex}


\begin{filecontents}{\jobname.bib}
@thesis{Luptak2016thesis,
  author       = {Lupták, Dávid},
  title        = {Typesetting of Bibliography According to ISO 690 Norm},
  date         = {2016},
  howpublished = {online},
  type         = {bathesis},
  institution  = {Masaryk University, Faculty of Informatics},
  location     = {Brno},
  supervisor   = {Petr Sojka},
  url          = {https://is.muni.cz/th/422640/fi_b/},
  urldate      = {2016-06-14},
}
\end{filecontents}
\addbibresource{\jobname.bib}


\begin{document}
\autocite{Luptak2016thesis}
\printbibliography
\end{document}
```

>  Brno [visited on 2016-06-14]. Bachelor’s thesis. Masaryk University, Faculty of Informatics. Supervised by Petr SOJKA. Available from: https://is.muni.cz/th/422640/fi_b/.

Is that really what ISO wants?

It is not as pronounced in
```latex
\documentclass[british]{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{babel}
\usepackage{csquotes}

\usepackage[style=iso-authoryear, thesisinfoinnotes=true, backend=biber]{biblatex}


\begin{filecontents}{\jobname.bib}
@thesis{Luptak2016thesis,
  author       = {Lupták, Dávid},
  title        = {Typesetting of Bibliography According to ISO 690 Norm},
  date         = {2016},
  howpublished = {online},
  type         = {bathesis},
  institution  = {Masaryk University, Faculty of Informatics},
  location     = {Brno},
  supervisor   = {Petr Sojka},
  url          = {https://is.muni.cz/th/422640/fi_b/},
  urldate      = {2016-06-14},
}
\end{filecontents}
\addbibresource{\jobname.bib}


\begin{document}
\autocite{Luptak2016thesis}
\printbibliography
\end{document}
```
but even there the punctuation seems a bit illogical

>  Brno [visited on 2016-06-14]. Available from: https://is.muni.cz/th/422640/fi_b/.

to me.

Is that what the ISO wants?

Reply:
Sorry, I forgot to comment on this. Unfortunately, this is what ISO wants. Especially for electronic documents, they specify it in pretty much detail:
```
For electronic monographs or contributions to monographs, the date when
   the material was accessed should follow:
a) the update/revision date, if one exists (see 8.2); or
b) if no update/revision date exists, the numeration/pagination, if either exist; or
c) if neither the update/revision date nor the numeration/pagination exist,
   the date of the publication (see 4.5 and 9.3).
```

Quite reasonable seems to print `urldate` together with `url` as the standard styles do in `url+urldate` bibmacro, but ISO 690 sticks to keep dates together rather than put date with other field that could relate to.

And to the `thesisinfoinnotes` option, I can recall that there is no strict position for this piece of information and many examples I found print it at different places, hence I created this option to adjust this behaviour if anybody would be interested. Now I am not able to find the respective thread where we were discussing it.

Reply:
I see, thanks for digging out the relevant rule. That appears to be incredibly counterintuitive, but if that's what ISO wants, what can I do?